File: mobile/ios/packages/town-crier-data/Sources/Auth/SessionCache.swift
import Foundation
import TownCrierDomain

/// Thread-safe in-memory cache for the current Auth0 session.
///
/// `Auth0AuthenticationService.currentSession()` was hitting
/// `SimpleKeychain.data(forKey:)` -> `SecItemCopyMatching` -> securityd on
/// every call, via `CredentialsManager.canRenew()` / `hasValid()` /
/// `credentials()`. Concurrent foreground bursts (e.g. a push tap arriving
/// while `LoginViewModel.checkExistingSession` and
/// `AppCoordinator.resolveSubscriptionTier` are already in flight) caused
/// four-way contention on securityd's connection pool, adding tens to
/// hundreds of milliseconds to cold-launch latency and amplifying the
/// resumption gap that drove the tc-cbmk notification-tap crash.
///
/// This cache holds the most recent valid `AuthSession` in memory. When the
/// session's access token has at least `leadTime` seconds before
/// `expiresAt`, callers get the cached value without touching the keychain.
/// Concurrent callers on a cold cache share a single in-flight load via
/// `currentOrLoad`, so a four-way burst issues at most one
/// `SecItemCopyMatching` (tc-3d7b).
actor SessionCache {
  private var cached: AuthSession?
  private var inFlight: Task<AuthSession?, Never>?
  private let leadTime: TimeInterval
  private let now: @Sendable () -> Date

  /// - Parameters:
  ///   - leadTime: How many seconds before `expiresAt` to consider the
  ///     cache stale. Mirrors Auth0's default 60-second renewal window so
  ///     we hand off to `CredentialsManager` before the access token
  ///     genuinely expires.
  ///   - now: Clock override for tests.
  init(
    leadTime: TimeInterval = 60,
    now: @Sendable @escaping () -> Date = { Date() }
  ) {
    self.leadTime = leadTime
    self.now = now
  }

  /// Returns a cached session if one is held and its access token is still
  /// at least `leadTime` seconds before `expiresAt`. Otherwise returns nil
  /// without populating the cache — callers should use `currentOrLoad` to
  /// fetch fresh credentials.
  func current() -> AuthSession? {
    guard let cached else { return nil }
    if cached.expiresAt.addingTimeInterval(-leadTime) > now() {
      return cached
    }
    return nil
  }

  /// Returns the cached session if valid, otherwise invokes `loader` and
  /// caches the result. Concurrent callers on a cold cache share the same
  /// in-flight load — this is the single-flight behaviour that prevents
  /// a foreground burst from issuing N keychain reads for the same token.
  func currentOrLoad(
    _ loader: @Sendable @escaping () async -> AuthSession?
  ) async -> AuthSession? {
    if let valid = current() {
      return valid
    }
    if let inFlight {
      return await inFlight.value
    }
    let task = Task<AuthSession?, Never> {
      await loader()
    }
    inFlight = task
    let result = await task.value
    cached = result
    inFlight = nil
    return result
  }

  /// Stores `session` as the current cached value. Used by `login` and
  /// `refreshSession` after a successful network round-trip so the next
  /// `currentSession()` short-circuits the keychain.
  func store(_ session: AuthSession) {
    cached = session
  }

  /// Drops the cached session and cancels any in-flight load. Called from
  /// `logout`, `deleteAccount`, and unrecoverable refresh failures.
  func clear() {
    cached = nil
    inFlight?.cancel()
    inFlight = nil
  }
}

File: mobile/android/data/src/main/kotlin/uk/towncrierapp/data/auth/SessionCache.kt
package uk.towncrierapp.data.auth

import uk.towncrierapp.domain.auth.AuthSession
import java.time.Clock
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * In-memory cache for the current [AuthSession] (iOS tc-3d7b). Concurrent
 * foreground bursts (a push tap arriving while the login check and the tier
 * resolver are already in flight) were each hitting
 * `SecureCredentialsManager` independently; this cache holds the most
 * recently loaded session while its access token has at least
 * [leadTimeSeconds] left before expiry, and single-flights concurrent cold
 * reads to one [loader] invocation via a shared [Deferred] so a four-way
 * burst issues at most one credentials-store read.
 *
 * [scope] is the rare case of an explicit, injected `CoroutineScope` for
 * process-lifetime work (android-coding-standards: coroutines-and-flow.md)
 * — the in-flight load must outlive whichever individual caller's coroutine
 * happens to create it, so every concurrent caller observes the same result.
 */
public class SessionCache(
    private val scope: CoroutineScope,
    private val leadTimeSeconds: Long = 60,
) {
    private val mutex = Mutex()
    private var cached: AuthSession? = null
    private var inFlight: Deferred<AuthSession?>? = null

    /** A cached session, or `null` if none is held or it's within [leadTimeSeconds] of expiry. Never triggers a load. */
    public suspend fun current(clock: Clock): AuthSession? = mutex.withLock { validOrNull(clock) }

    private fun validOrNull(clock: Clock): AuthSession? = cached?.takeIf { it.expiresAt.minusSeconds(leadTimeSeconds).isAfter(clock.instant()) }

    /** Returns the cached session if valid, otherwise runs [loader] — sharing one in-flight call across concurrent cold callers. */
    public suspend fun currentOrLoad(
        clock: Clock,
        loader: suspend () -> AuthSession?,
    ): AuthSession? {
        val deferred =
            mutex.withLock {
                validOrNull(clock)?.let { return it }
                inFlight ?: scope.async { loader() }.also { inFlight = it }
            }
        val result = deferred.await()
        mutex.withLock {
            if (inFlight === deferred) {
                cached = result
                inFlight = null
            }
        }
        return result
    }

    /** Stores [session] directly — used after a successful login or refresh round-trip. */
    public suspend fun store(session: AuthSession) {
        mutex.withLock { cached = session }
    }

    /** Drops the cached session and cancels any in-flight load — used on logout and unrecoverable refresh failures. */
    public suspend fun clear() {
        mutex.withLock {
            cached = null
            inFlight?.cancel()
            inFlight = null
        }
    }
}
